text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,581 @@
+of new inline _type unions_. Exceptions will be added for all existing APIs.
+
+## Generating _union-like_ types
+
+Code generation needs to be changed in `jsii-pacmak` so that languages without native support for _type unions_ (**C#**, **Go**, and **Java**) have
+additional _union-like_ types emitted. The proposed API for a union such as `export type UnionType = Foo | Map<Bar> | Baz[]` in each currently
+supported statically typed language is the following (subject ot change):
+
+- In **C#**
+
+  ```csharp
+  // UnionType.cs
+  public sealed class UnionType {
+    public static UnionType FromFoo(Foo value) { /* ... */ }
+    public static UnionType FromMapOfBar(IReadOnlyDictionary<Bar> value) { /* ... */ }
+    public static UnionType FromListOfBaz(IReadOnlyList<Baz> value) { /* ... */ }
+
+    public Foo? Foo { get; }
+    public IReadOnlyDictionary<Bar>? MapOfBar { get; }
+    public IReadOnlyList<Baz>? ListOfBaz { get; }
+
+    // Private constructor ensures all initialization happens though the factories
+    private UnionType(/*...*/) { /*...*/ }
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public abstract class UnionType {
+    public static UnionType fromFoo(Foo value) {
+      return new UnionType.FooUnionType(value);
+    }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) {
+      return new UnionType.MapOfBarUnionType(value);
+    }
+    public static UnionType fromListOfBaz(Baz... value) {
+      return fromListOfBaz(Arrays.asList(value));
+    }
+    public static UnionType fromListOfBaz(List<Baz> value) {
+      return new UnionType.ListOfBaz(value);
+    }
+
+    public @Nullable Foo asFoo() { return null; }
+    public @Nullable Map<String, Bar> asMapOfBar() { return null; }
+    public @Nullable List<Baz> asListOfBaz() { return null; }
+
+    // Private constructor ensures all initialization happens through the factories
+    private UnionType(/* ... */) { /* ... */ }
+
+    private static final class FooUnionType extends UnionType { /*...*/ }
+    private static final class MapOfBarUnionType extends UnionType { /*...*/ }
+    private static final class ListOfBazUnionType extends UnionType { /*...*/ }
+  }
+  ```
+
+The _jsii Kernel_ will need adjustments in order to better support those types. The generated libraries will provide a new argument to the _jsii
+kernel_'s `load` API if it's code was generated with `--union-like-types` enabled. This way, the kernel will be bale to reference the declared _type
+union_'s information when annotating instances that are passed from **Javascript** to the host application, removing all ambiguity regarding which
+type is serialized.
+
+# Drawbacks
+
+This approach may allow developers in statically typed languages to cause _undefined behavior_ to occur if they are not provided with a way to
```

Comment:
Wait what? I don't understand this entire section.

> if they are not provided with a way to validate

In statically typed languages that use our generated union class, we can GIVE them a way to validate the type, no? We can add an `isFoo(): boolean` method.

Further, you initially defined `asFoo()` to return `null` if the enclosed type wasn't a `Foo`, so that could serve as a type check (albeit one I disagree with).

> developers may use an invalid `as<TypeName>` accessor and retrive a proxy that may behave unexpectedly

You defined it to return `null`, which is not a proxy at all. If we throw an exception instead, we don't return a proxy at all, which is also not a problem.

> This behavior is however identical to what happens in **TypeScript** when developers use `as any`

It is not? In TypeScript you'll get "'8' does not have a method 'toLowerCase()'", whereas this would (as currently specced) return `null`. 

-----

I think you might either have meant to say "dynamically typed languages" or "statically typed languages with union type support so that we don't use our generated union class" ? Otherwise I don't really understand what you're trying to say here.

Reply:
My argument is we may not always be able to type validate completely (e.g: I cannot tell two homonymous methods apart at runtime). In such cases, runtime could allow a forceful downcast that is technically incorrect, and you're in the same situation as if you had used `as any` in TypeScript.

Of course, if the API includes a type assertion method, this can be used to confirm whether your cast is valid or not (either before or after having done it, I suppose).

---

> It is not? In TypeScript you'll get "'8' does not have a method 'toLowerCase()'", whereas this would (as currently specced) return null.

I mean identical to what happens if you use `as any` in TypeScript in the sense that you could get an error if you're attempting to do things to this value that it does not support/allow. The error could be different (but it'd be an error nonetheless).

Also - if we craft an "arbitrary type" proxy for this, we'd forward the call straight to the JS process, who'd then fail it the same way as it would in TypeScript.